bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -788,89 +852,340 @@ private long readLastLogId(File f) {
+         * not existing.
+         */
+        BufferedLogChannel getCurrentLogIfPresent(long entryLogId);
+
+        /*
+         * Returns eligible writable ledger dir for the creation next entrylog
+         */
+        File getDirForNextEntryLog(List<File> writableLedgerDirs);
+
+        /*
+         * Do the operations required for checkpoint.
+         */
+        void checkpoint() throws IOException;
+
+        /*
+         * roll entryLogs.
+         */
+        void rollLogs() throws IOException;
+
+        /*
+         * flush rotated logs.
+         */
+        void flushRotatedLogs() throws IOException;
+
+        /*
+         * flush current logs.
+         */
+        void flushCurrentLogs() throws IOException;
+
+        /*
+         * close current logs.
+         */
+        void closeCurrentLogs() throws IOException;
+
+        /*
+         * force close current logs.
+         */
+        void forceCloseCurrentLogs();
+
+        /*
+         * this method should be called before doing entrymemtable flush, it
+         * would save the state of the entrylogger before entrymemtable flush
+         * and commitEntryMemTableFlush would take appropriate action after
+         * entrymemtable flush.
+         */
+        void prepareEntryMemTableFlush();
+
         /*
-         * In the case of entryLogPerLedgerEnabled we need to flush both
-         * rotatedlogs and currentlogs. This is needed because syncThread
-         * periodically does checkpoint and at this time all the logs should
-         * be flushed.
+         * this method should be called after doing entrymemtable flush,it would
+         * take appropriate action after entrymemtable flush depending on the
+         * current state of the entrylogger and the state of the entrylogger
+         * during prepareEntryMemTableFlush.
          *
-         * TODO: When EntryLogManager is introduced in the subsequent sub-tasks of
-         * this Issue, I will move this logic to individual implamentations of
-         * EntryLogManager and it would be free of this booalen flag based logic.
+         * It is assumed that there would be corresponding
+         * prepareEntryMemTableFlush for every commitEntryMemTableFlush and both
+         * would be called from the same thread.
          *
+         * returns boolean value indicating whether EntryMemTable should do checkpoint
+         * after this commit method.
          */
-        if (entryLogPerLedgerEnabled) {
-            flushCurrentLog();
-        }
+        boolean commitEntryMemTableFlush() throws IOException;
     }
 
-    void flushRotatedLogs() throws IOException {
-        List<BufferedLogChannel> channels = null;
-        long flushedLogId = INVALID_LID;
-        synchronized (this) {
-            channels = logChannelsToFlush;
-            logChannelsToFlush = null;
+    abstract class EntryLogManagerBase implements EntryLogManager {
+        final Set<BufferedLogChannel> rotatedLogChannels;
```

Comment:
1) in both the cases (singleentrylog and entrylogperledger) rotatedLogChannels are dealt similarly. So it makes sense to have this variable and related methods (flushRotatedLogs) in Base class.
2) I dont see it to be much of an issue. Anyhow, I'll change it to List interface, but it is incorrect to have non-synchronized version of List for rotatedLogChannels. I'll use CopyOnWriteArrayList, so that the list can be modified while its iterator is being used.

Reply:
>  it is incorrect to have non-synchronized version of List for rotatedLogChannels. I'll use CopyOnWriteArrayList, so that the list can be modified while its iterator is being used.

why? the original code is synchronizing the rotatedLogChannels list properly. If you just move the logic into the abstraction, this logic should work.